File: app/src/main/java/ch/bailu/aat/services/sensor/attributes/CadenceSpeedAttributes.java
package ch.bailu.aat.services.sensor.attributes;

public class CadenceSpeedAttributes   extends IndexedAttributes {


    public final static String[] SENSOR_LOCATION = {
            "Other",
            "Top of shoe",
            "In shoe",
            "Hip",
            "Front Wheel",
            "Left Crank",
            "Right Crank",
            "Left Pedal",
            "Right Pedal",
            "Front Hub",
            "Rear Dropout",
            "Chainstay",
            "Rear Wheel",
            "Rear Hub",
            "Chest",
            "Spider",
            "Chain Ring",
    };


    public static final int KEY_INDEX_SENSOR_LOCATION = 0;
    public static final int KEY_INDEX_SPEED_SENSOR = 1;
    public static final int KEY_INDEX_CADENCE_SENSOR = 2;
    public static final int KEY_INDEX_CRANK_RPM = 3;
    public static final int KEY_INDEX_WHEEL_CIRCUMFERENCE = 4;
    public static final int KEY_INDEX_CONTACT = 5;


    public final static String[] KEYS = {
            "Location",
            "SpeedSensor",
            "CadenceSensor",
            "Cadence",
            "WheelCircumference",
            "Contact"
    };


    public int cadence_rpm = 0;
    public int cadence_rpm_average = 0;
    public float circumferenceSI = 0f;

    public final String location;
    public final boolean isCadenceSensor, isSpeedSensor;


    public CadenceSpeedAttributes(String l, boolean cadence, boolean speed) {
        super(KEYS);
        location = l;
        isCadenceSensor = cadence;
        isSpeedSensor = speed;
    }



    @Override
    public String getValue(int index) {
        if (index == KEY_INDEX_SENSOR_LOCATION) {
            return location;

        } else if (index == KEY_INDEX_CADENCE_SENSOR) {
            return String.valueOf(isCadenceSensor);

        } else if (index == KEY_INDEX_SPEED_SENSOR) {
            return String.valueOf(isSpeedSensor);

        } else if (index == KEY_INDEX_CRANK_RPM) {
            return String.valueOf(cadence_rpm_average);

        } else if (index == KEY_INDEX_WHEEL_CIRCUMFERENCE) {
            return String.valueOf(circumferenceSI);

        } else if (index == KEY_INDEX_CONTACT) {
            if (cadence_rpm == 0) return "...";
            return "";
        }


        return NULL_VALUE;
    }
}

File: aat-lib/src/main/java/ch/bailu/aat_lib/gpx/attributes/CadenceSpeedAttributes.kt
package ch.bailu.aat_lib.gpx.attributes

open class CadenceSpeedAttributes(
    val location: String,
    private val isCadenceSensor: Boolean,
    private val isSpeedSensor: Boolean
) : GpxAttributes() {
    var cadenceRpm = 0
    var cadenceRpmAverage = 0
    var circumferenceSI = 0f
    var circumferenceDebugString = "--"

    override fun get(keyIndex: Int): String {
        if (keyIndex == KEY_INDEX_SENSOR_LOCATION) {
            return location
        } else if (keyIndex == KEY_INDEX_CADENCE_SENSOR) {
            return isCadenceSensor.toString()
        } else if (keyIndex == KEY_INDEX_SPEED_SENSOR) {
            return isSpeedSensor.toString()
        } else if (keyIndex == KEY_INDEX_CRANK_RPM) {
            return cadenceRpmAverage.toString()
        } else if (keyIndex == KEY_INDEX_WHEEL_CIRCUMFERENCE) {
            return circumferenceSI.toString()
        } else if (keyIndex == KEY_INDEX_CONTACT) {
            return getAsBoolean(keyIndex).toString()
        } else if (keyIndex == KEY_INDEX_CIRCUMFERENCE_DEBUG) {
            return circumferenceDebugString
        }
        return NULL_VALUE
    }

    override fun getAsBoolean(keyIndex: Int): Boolean {
        return if (keyIndex == KEY_INDEX_CONTACT) {
            val cadenceContact = isCadenceSensor && cadenceRpm > 0
            val speedContact = isSpeedSensor && circumferenceSI > 0f
            cadenceContact || speedContact
        } else {
            super.getAsBoolean(keyIndex)
        }
    }

    override fun hasKey(keyIndex: Int): Boolean {
        return KEYS.hasKey(keyIndex)
    }

    override fun size(): Int {
        return KEYS.size()
    }

    override fun getAt(index: Int): String {
        return get(KEYS.getKeyIndex(index))
    }

    override fun getKeyAt(index: Int): Int {
        return KEYS.getKeyIndex(index)
    }

    companion object {
        @JvmField
        val SENSOR_LOCATION = arrayOf(
            "Other",
            "Top of shoe",
            "In shoe",
            "Hip",
            "Front Wheel",
            "Left Crank",
            "Right Crank",
            "Left Pedal",
            "Right Pedal",
            "Front Hub",
            "Rear Dropout",
            "Chainstay",
            "Rear Wheel",
            "Rear Hub",
            "Chest",
            "Spider",
            "Chain Ring"
        )
        @JvmField
        val KEYS = Keys()
        val KEY_INDEX_SENSOR_LOCATION = KEYS.add("Location")
        val KEY_INDEX_SPEED_SENSOR = KEYS.add("SpeedSensor")
        val KEY_INDEX_CADENCE_SENSOR = KEYS.add("CadenceSensor")
        @JvmField
        val KEY_INDEX_CRANK_RPM = KEYS.add("Cadence")
        val KEY_INDEX_WHEEL_CIRCUMFERENCE = KEYS.add("WheelCircumference")
        @JvmField
        val KEY_INDEX_CONTACT = KEYS.add("Contact")
        val KEY_INDEX_CIRCUMFERENCE_DEBUG = KEYS.add("Debug")
    }
}
